autogen/code_utils.py

```diff
@@ -229,20 +229,28 @@ def get_powershell_command():
                 return "pwsh"
 
         except (FileNotFoundError, NotADirectoryError):
-            if WIN32:
-                logging.warning("Neither powershell nor pwsh is installed but it is a Windows OS")
-            return None
+            if WIN32 and FileNotFoundError:
+                logging.warning("Neither powershell.exe nor pwsh.exe is present but it is a Windows OS")
+
+            elif WIN32 and NotADirectoryError:
+                logging.warning(
+                    "PowerShell is either not installed or its path is not given properly in the environment variable PATH. Please check the path and try again."
+                )
 
+            return None
 
-powershell_command = get_powershell_command()
+    except PermissionError:
+        logging.warning("The application has no permission to run powershell")
+        return None
 
 
 def _cmd(lang):
-    if lang.startswith("python") or lang in ["bash", "sh", powershell_command]:
+    if lang.startswith("python") or lang in ["bash", "sh"]:
         return lang
     if lang in ["shell"]:
         return "sh"
     if lang in ["ps1", "pwsh", "powershell"]:
+        powershell_command = get_powershell_command()
```

Comment:
Because get_powershell_command may return None, need to handle the None case here.

Reply:
In my opinion, it is a hard error. You requested poweshell and it is not available.

Reply:
How would you suggest we solve this?

Reply:
I think we should throw an exception

Reply:
yeah, the line below shouldn't return 'sh', that would be unexpected. So I think we should throw an exception inside the `get_powershell_command` function instead of logging warnings.

Reply:
And we should also include the original exception in the error message.

Reply:
I changed the logging.warnings to raise exceptions now where needed. And I've made the get_powershell_command function return "sh" in the except blocks if it is not a windows OS (WIN32) since the bash terminal is the most commonly used terminal outside of powershell, is that okay?

Reply:
If the codeblock contains powershell code, and we cannot get powershell then it is an error. There's no guarnatee `sh` is going to work. Maybe getting powershell or maybe getting shell makes the function `get_powershell_command` confusing and hard to use

Reply:
Okay yes that makes sense I'll just make it raise an exception then instead of returning sh